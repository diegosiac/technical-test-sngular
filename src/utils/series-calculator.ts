export class SeriesCalculator {
  private static validateInput(number: bigint): void {
    if (typeof number !== 'bigint' || number <= 0n) {
      throw new RangeError('El número debe ser un número natural (no negativo)')
    }
  }

  static triangular(number: bigint): bigint {
    SeriesCalculator.validateInput(number)

    return (number * (number + 1n)) / 2n
  }

  static fibonacci(number: bigint): bigint {
    SeriesCalculator.validateInput(number)

    if (number === 1n || number === 2n) return 1n

    let a = 1n
    let b = 1n

    for (let i = 3n; i <= number; i++) {
      [a, b] = [b, a + b]
    }

    return b
  }

  private static isPrime(number: bigint): boolean {
    if (number < 2n) return false

    for (let i = 2n; i * i <= number; i++) {
      if (number % i === 0n) return false
    }

    return true
  }

  static prime(number: bigint): bigint {
    SeriesCalculator.validateInput(number)

    let count = 0n
    let current = 2n

    while (true) {
      if (SeriesCalculator.isPrime(current)) {
        count++
      }

      if (count === number) {
        return current
      }

      current++
    }
  }

  static computeTerm(value: number): bigint {
    const number = BigInt(value)
    SeriesCalculator.validateInput(number)

    const triangular = SeriesCalculator.triangular(number * 2n)
    const prime = SeriesCalculator.prime(number)
    const fibonacci = SeriesCalculator.fibonacci(number)

    return triangular - prime - fibonacci
  }
}
