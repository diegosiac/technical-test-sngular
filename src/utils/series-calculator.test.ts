import { describe, expect, it, test } from 'vitest'
import { SeriesCalculator } from './series-calculator'

describe('SeriesCalculator', () => {
  const testErrorCases = (method: (input: bigint) => bigint) => {
    describe('should throw RangeError for', () => {
      it('undefined input', () => {
        expect(() => method(undefined as any)).toThrowErrorMatchingSnapshot()
      })

      it('negative numbers', () => {
        expect(() => method(-5n)).toThrowErrorMatchingSnapshot()
      })

      it('zero', () => {
        expect(() => method(0n)).toThrowErrorMatchingSnapshot()
      })
    })
  }

  describe('triangular() method', () => {
    testErrorCases(SeriesCalculator.triangular)

    describe('valid inputs', () => {
      const testCases = [
        { input: 1n, expected: 1n },
        { input: 3n, expected: 6n },
        { input: 10n, expected: 55n },
        { input: 100n, expected: 5050n },
      ]

      test.each(testCases)('$input => $expected', ({ input, expected }) => {
        expect(SeriesCalculator.triangular(input)).toBe(expected)
      })
    })
  })

  describe('fibonacci() method', () => {
    testErrorCases(SeriesCalculator.fibonacci)

    describe('valid inputs', () => {
      const testCases = [
        { input: 1n, expected: 1n },
        { input: 2n, expected: 1n },
        { input: 10n, expected: 55n },
        { input: 20n, expected: 6765n },
        { input: 50n, expected: 12586269025n },
      ]

      test.each(testCases)('$input => $expected', ({ input, expected }) => {
        expect(SeriesCalculator.fibonacci(input)).toBe(expected)
      })
    })
  })

  describe('prime() method', () => {
    testErrorCases(SeriesCalculator.prime)

    describe('valid inputs', () => {
      const testCases = [
        { input: 1n, expected: 2n },
        { input: 4n, expected: 7n },
        { input: 10n, expected: 29n },
        { input: 50n, expected: 229n },
        { input: 100n, expected: 541n },
      ]

      test.each(testCases)('$input => $expected', ({ input, expected }) => {
        expect(SeriesCalculator.prime(input)).toBe(expected)
      })
    })
  })

  describe('computeTerm() method', () => {
    const testCases = [
      { input: 1, expected: 0n },
      { input: 2, expected: 6n },
      { input: 5, expected: 39n },
      { input: 10, expected: 126n },
      { input: 13, expected: 77n },
      { input: 17, expected: -1061n },
      { input: 20, expected: -6016n },
      { input: 25, expected: -73847n },
      { input: 30, expected: -830323n },
      {input: 50, expected: -12586264204n},
    ]

    test.each(testCases)('term $input => $expected', ({ input, expected }) => {
      expect(SeriesCalculator.computeTerm(input)).toBe(expected)
    })

    test('should throw for invalid inputs', () => {
      expect(() => SeriesCalculator.computeTerm(0)).toThrow()
      expect(() => SeriesCalculator.computeTerm(-1)).toThrow()
      expect(() => SeriesCalculator.computeTerm(1.5)).toThrow()
    })
  })
})
