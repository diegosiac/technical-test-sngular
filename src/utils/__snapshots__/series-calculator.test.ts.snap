// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SeriesCalculator > fibonacci() method > should throw RangeError for > negative numbers 1`] = `[RangeError: El número debe ser un número natural (no negativo)]`;

exports[`SeriesCalculator > fibonacci() method > should throw RangeError for > undefined input 1`] = `[RangeError: El número debe ser un número natural (no negativo)]`;

exports[`SeriesCalculator > fibonacci() method > should throw RangeError for > zero 1`] = `[RangeError: El número debe ser un número natural (no negativo)]`;

exports[`SeriesCalculator > prime() method > should throw RangeError for > negative numbers 1`] = `[RangeError: El número debe ser un número natural (no negativo)]`;

exports[`SeriesCalculator > prime() method > should throw RangeError for > undefined input 1`] = `[RangeError: El número debe ser un número natural (no negativo)]`;

exports[`SeriesCalculator > prime() method > should throw RangeError for > zero 1`] = `[RangeError: El número debe ser un número natural (no negativo)]`;

exports[`SeriesCalculator > triangular() method > should throw RangeError for > negative numbers 1`] = `[RangeError: El número debe ser un número natural (no negativo)]`;

exports[`SeriesCalculator > triangular() method > should throw RangeError for > undefined input 1`] = `[RangeError: El número debe ser un número natural (no negativo)]`;

exports[`SeriesCalculator > triangular() method > should throw RangeError for > zero 1`] = `[RangeError: El número debe ser un número natural (no negativo)]`;
